gobblin-core/src/main/java/gobblin/source/extractor/extract/google/GoogleAnalyticsUnsampledExtractor.java

```diff
@@ -0,0 +1,367 @@
+    try {
+      return retryer.call(new Callable<UnsampledReport>() {
+        @Override
+        public UnsampledReport call() throws Exception {
+          UnsampledReport response = insertRequest.execute();
+          if (FAILED_STATUS.equals(response.getStatus())) { //No retry if it's explicitly failed from server
+            throw new NonTransientException("Failed to create unsampled report " + response);
+          }
+          response.setEndDate(endDate); //response does not have end date where we need it later for next watermark calculation.
+          return response;
+        }
+      });
+    } catch (ExecutionException e) {
+      throw new IOException(e);
+    } catch (RetryException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Converts date format from watermark format to Google analytics format
+   * @param watermark
+   * @return
+   */
+  private String convertFormat(long watermark) {
+    Preconditions.checkArgument(watermark > 0, "Watermark should be positive number.");
+    return googleAnalyticsFormatter.print(watermarkFormatter.parseDateTime(Long.toString(watermark)));
+  }
+
+  @VisibleForTesting
+  UnsampledReport pollForCompletion(State state, final Analytics gaService, final UnsampledReport requestedReport) throws IOException {
+
+    Config config = ConfigBuilder.create()
+                                 .loadProps(state.getProperties(), POLL_RETRY_PREFIX)
+                                 .build()
+                                 .withFallback(POLL_RETRY_DEFAULTS);
+
+    Retryer<UnsampledReport> retryer = RetryerFactory.newInstance(config);
+    LOG.info("Will poll for completion on unsampled report with retry config: " + config);
+
+    final Stopwatch stopwatch = Stopwatch.createStarted();
+    UnsampledReport result = null;
+    try {
+      result = retryer.call(new Callable<UnsampledReport>() {
+
+        @Override
+        public UnsampledReport call() throws Exception {
+          UnsampledReport response = null;
+          try {
+            response = gaService.management()
+                .unsampledReports()
+                .get(requestedReport.getAccountId(),
+                     requestedReport.getWebPropertyId(),
+                     requestedReport.getProfileId(),
+                     requestedReport.getId())
+                .execute();
+          } catch (Exception e) {
+            LOG.warn("Encountered exception while polling for unsampled report. Will keep polling. " +
+                     "Elasped so far: " + stopwatch.elapsed(TimeUnit.SECONDS) + " seconds", e);
+            throw e;
+          }
+
+          if (FAILED_STATUS.equals(response.getStatus())) {
+            //Stop retrying if it explicitly failed from server.
+            throw new NonTransientException("Unsampled report has failed to be generated. " + response);
+          }
+
+          if (!COMPLETED_STATUS.equals(response.getStatus())) {
```

Comment:
I think it might be a better idea to throw on PENDING status rather than not-COMPLETED status, because in future if the API extends to throw say ABORTED or TIMEDOUT instead of FAILED status, we will un-necessarily continue to poll. Similarly it would be a good idea to ensure that COMPLETED status is returned instead of implicitly assuming it that it is completed.